From 50193efba654630e24547e9acf154822cbd612b8 Mon Sep 17 00:00:00 2001
From: wanghang73 <wanghang73@huawei.com>
Date: Tue, 28 Apr 2026 22:15:12 +0800
Subject: [PATCH] urma: bonding device should not use wr entry if jetty deleted

- Revert "urma: fix port down recover error bug"
  This reverts commit 20cd28aa4e0d5b93642922c1521a2f576921545d.
- fix bonding device wr buf clear
---
 src/urma/lib/urma/bond/bondp_api.c      |  8 +++++
 src/urma/lib/urma/bond/bondp_datapath.c | 42 ++++++++++++++++---------
 src/urma/lib/urma/bond/bondp_types.h    |  1 +
 src/urma/lib/urma/bond/bondp_wr_buf.c   |  1 +
 4 files changed, 37 insertions(+), 15 deletions(-)

diff --git a/src/urma/lib/urma/bond/bondp_api.c b/src/urma/lib/urma/bond/bondp_api.c
index b06e83a6..2be0cbfc 100644
--- a/src/urma/lib/urma/bond/bondp_api.c
+++ b/src/urma/lib/urma/bond/bondp_api.c
@@ -629,6 +629,7 @@ urma_jfs_t *bondp_create_jfs(urma_context_t *ctx, urma_jfs_cfg_t *cfg)
     (void)pthread_spin_init(&bdp_jfs->send_lock, PTHREAD_PROCESS_PRIVATE);
     bdp_jfs->send_jfc = CONTAINER_OF_FIELD(cfg->jfc, bondp_jfc_t, v_jfc);
     bdp_jfs->recv_jfc = NULL;
+    bdp_jfs->modify_to_error = false;
     for (uint32_t i = 0; i < URMA_UBAGG_DEV_MAX_NUM; i++) {
         atomic_init(&bdp_jfs->sqe_cnt[i], 0);
     }
@@ -733,6 +734,9 @@ urma_status_t bondp_modify_jfs(urma_jfs_t *jfs, urma_jfs_attr_t *attr)
     urma_status_t ret = URMA_SUCCESS, final_ret = URMA_SUCCESS;
     bondp_comp_t *bdp_jfs = CONTAINER_OF_FIELD(jfs, bondp_comp_t, v_jfs);
 
+    if ((attr->mask & JFS_STATE) && attr->state == URMA_JETTY_STATE_ERROR) {
+        bdp_jfs->modify_to_error = true;
+    }
     for (int i = 0; i < URMA_UBAGG_DEV_MAX_NUM; i++) {
         if (bdp_jfs->p_jfs[i] == NULL) {
             continue;
@@ -1210,6 +1214,7 @@ urma_jetty_t *bondp_create_jetty(urma_context_t *ctx, urma_jetty_cfg_t *jetty_cf
     (void)pthread_spin_init(&bdp_jetty->send_lock, PTHREAD_PROCESS_PRIVATE);
     bdp_jetty->send_jfc = CONTAINER_OF_FIELD(jetty_cfg->jfs_cfg.jfc, bondp_jfc_t, v_jfc);
     bdp_jetty->recv_jfc = CONTAINER_OF_FIELD(jetty_cfg->shared.jfr->jfr_cfg.jfc, bondp_jfc_t, v_jfc);
+    bdp_jetty->modify_to_error = false;
 
     const bondp_port_id_t *cfg_active_port_ids = NULL;
     uint32_t cfg_active_port_count = 0;
@@ -1330,6 +1335,9 @@ urma_status_t bondp_modify_jetty(urma_jetty_t *jetty, urma_jetty_attr_t *attr)
     urma_status_t ret = URMA_SUCCESS, final_ret = URMA_SUCCESS;
     bondp_comp_t *bdp_jetty = CONTAINER_OF_FIELD(jetty, bondp_comp_t, v_jetty);
 
+    if ((attr->mask & JETTY_STATE) && attr->state == URMA_JETTY_STATE_ERROR) {
+        bdp_jetty->modify_to_error = true;
+    }
     for (int i = 0; i < URMA_UBAGG_DEV_MAX_NUM; i++) {
         if (bdp_jetty->p_jetty[i] == NULL) {
             continue;
diff --git a/src/urma/lib/urma/bond/bondp_datapath.c b/src/urma/lib/urma/bond/bondp_datapath.c
index 152e76b5..b4070375 100644
--- a/src/urma/lib/urma/bond/bondp_datapath.c
+++ b/src/urma/lib/urma/bond/bondp_datapath.c
@@ -305,7 +305,6 @@ static urma_status_t bondp_post_send_wr_and_store(bondp_comp_t *bdp_comp, urma_j
         ret = schedule_send(wr->tjetty, bdp_comp, &send_idx, &target_idx, &info);
     }
     if (ret != 0) {
-        URMA_LOG_WARN("Failed to schedule send.\n");
         return URMA_FAIL;
     }
 
@@ -349,7 +348,9 @@ RELEASE_WR:
     release_vwr_use_cnt(pwr);
 FREE_PWR:
     free_jfs_wr(pwr);
+    (void)pthread_spin_lock(&bdp_comp->send_jfc->wr_lock);
     jfs_wr_buf_release(wr_entry);
+    (void)pthread_spin_unlock(&bdp_comp->send_jfc->wr_lock);
     return ret;
 }
 
@@ -654,7 +655,7 @@ typedef enum cr_convert_ret {
     CONVERT_SKIP = 1,
 } cr_convert_ret_t;
 
-static int resend_jfs_wr(jfs_wr_entry_t *wr_entry, int send_idx, int target_idx)
+static int resend_jfs_wr(bondp_comp_t *bdp_comp, jfs_wr_entry_t *wr_entry, int send_idx, int target_idx)
 {
     wr_entry->send_idx = send_idx;
     wr_entry->target_idx = target_idx;
@@ -670,7 +671,9 @@ static int resend_jfs_wr(jfs_wr_entry_t *wr_entry, int send_idx, int target_idx)
         convert_jfs_pwr_to_vwr_resend(wr, vtjetty);
         release_vwr_use_cnt(wr);
         free_jfs_wr(wr);
+        (void)pthread_spin_lock(&bdp_comp->send_jfc->wr_lock);
         jfs_wr_buf_release(wr_entry);
+        (void)pthread_spin_unlock(&bdp_comp->send_jfc->wr_lock);
     }
 
     return ret;
@@ -754,7 +757,7 @@ static cr_convert_ret_t handle_fake_cr_with_store(bondp_context_t *bdp_ctx, int
     return CONVERT_SKIP;
 }
 
-static cr_convert_ret_t handle_send_cr_with_store(bondp_jfc_t *bdp_jfc, int idx, urma_cr_t *cr)
+static cr_convert_ret_t handle_send_cr_with_store(bondp_context_t *bdp_ctx, bondp_jfc_t *bdp_jfc, int idx, urma_cr_t *cr)
 {
     const uint64_t wr_id = cr->user_ctx;
     jfs_wr_entry_t *wr_entry = jfs_wr_buf_get(&bdp_jfc->wr_buf, wr_id);
@@ -769,15 +772,26 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_jfc_t *bdp_jfc, int idx,
         return CONVERT_SKIP;
     }
 
-    bondp_comp_t *bdp_comp = wr_entry->bdp_comp;
+    bondp_comp_t *bdp_comp = get_comp_by_cr(bdp_ctx, idx, cr);
+    if (bdp_comp == NULL) {
+        URMA_LOG_ERR("Failed find jetty when handle send cr, cr.local_id:%u.\n", cr->local_id);
+        return CONVERT_SKIP;
+    }
+    if (wr_entry->bdp_comp != bdp_comp) {
+        URMA_LOG_ERR("Corrupted wr entry, bdp_comp not match, in_entry:%p, found:%p.\n", wr_entry->bdp_comp, bdp_comp);
+        put_comp(bdp_comp);
+        return CONVERT_SKIP;
+    }
+
     uint32_t send_idx = wr_entry->send_idx;
     uint32_t target_idx = wr_entry->target_idx;
 
     if (bdp_comp->valid[idx] == false || idx != send_idx) {
+        put_comp(bdp_comp);
         return CONVERT_SKIP;
     }
 
-    if (is_failover_cr(cr)) {
+    if (is_failover_cr(cr) && !bdp_comp->modify_to_error) {
         (void)pthread_spin_lock(&bdp_comp->send_lock);
         bdp_comp->valid[send_idx] = false;
 
@@ -789,17 +803,9 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_jfc_t *bdp_jfc, int idx,
              * this error CQE is returned directly to the upper layer.
              */
             URMA_LOG_ERR("Failed to find valid port for retransmission.\n");
-            /* Port down can be recovered, currently for simplex mode */
-            if (cr->status == URMA_CR_LOC_ACCESS_ERR && bdp_comp->comp_type == BONDP_COMP_JFS) {
-                bdp_comp->valid[send_idx] = true;
-            }
             (void)pthread_spin_unlock(&bdp_comp->send_lock);
             goto CONVERT_CR;
         }
-        /* Port down can be recovered, currently for simplex mode */
-        if (cr->status == URMA_CR_LOC_ACCESS_ERR && bdp_comp->comp_type == BONDP_COMP_JFS) {
-            bdp_comp->valid[send_idx] = true;
-        }
 
         URMA_LOG_DEBUG("Resend from %d to %d\n", send_idx, new_send_idx);
 
@@ -807,12 +813,14 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_jfc_t *bdp_jfc, int idx,
             const uint64_t wr_id = (wr_entry->wr_id + i - 1) % bdp_jfc->wr_buf.max_wr_num + 1;
             jfs_wr_entry_t *resend_wr_entry = jfs_wr_buf_get(&bdp_jfc->wr_buf, wr_id);
             if (resend_wr_entry == NULL ||
+                resend_wr_entry->entry_type != WR_BUF_ENTRY_JFS ||
+                resend_wr_entry->bdp_comp != bdp_comp ||
                 resend_wr_entry->send_idx != send_idx ||
                 resend_wr_entry->target_idx != target_idx) {
                 continue;
             }
             atomic_fetch_sub(&bdp_comp->sqe_cnt[send_idx], 1);
-            if (resend_jfs_wr(resend_wr_entry, new_send_idx, new_target_idx) != 0) {
+            if (resend_jfs_wr(bdp_comp, resend_wr_entry, new_send_idx, new_target_idx) != 0) {
                 URMA_LOG_ERR("Failed to resend jfs wr, wr_id: %lu\n", wr_id);
             }
         }
@@ -837,6 +845,7 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_jfc_t *bdp_jfc, int idx,
             bondp_notify_health_event(bdp_comp->bondp_ctx, BONDP_HEALTH_EVENT_FALLBACK_TASK_KICK, &event_info);
         }
         (void)pthread_spin_unlock(&bdp_comp->send_lock);
+        put_comp(bdp_comp);
         return CONVERT_SKIP;
     }
 
@@ -851,7 +860,10 @@ CONVERT_CR:
     convert_jfs_pwr_to_vwr_resend(&wr_entry->wr, &wr_entry->v_conn->target_vjetty->v_tjetty);
     release_vwr_use_cnt(&wr_entry->wr);
     free_jfs_wr(&wr_entry->wr);
+    (void)pthread_spin_lock(&bdp_comp->send_jfc->wr_lock);
     jfs_wr_buf_release(wr_entry);
+    (void)pthread_spin_unlock(&bdp_comp->send_jfc->wr_lock);
+    put_comp(bdp_comp);
     return CONVERT_SUCCESS;
 }
 
@@ -957,7 +969,7 @@ static cr_convert_ret_t bondp_handle_cr_with_store(bondp_context_t *bdp_ctx, bon
     } else if (is_recv_cr(cr)) {
         return handle_recv_cr_with_store(bdp_ctx, bdp_jfc, idx, cr);
     } else {
-        return handle_send_cr_with_store(bdp_jfc, idx, cr);
+        return handle_send_cr_with_store(bdp_ctx, bdp_jfc, idx, cr);
     }
 }
 
diff --git a/src/urma/lib/urma/bond/bondp_types.h b/src/urma/lib/urma/bond/bondp_types.h
index a7a0c53e..b91a0791 100644
--- a/src/urma/lib/urma/bond/bondp_types.h
+++ b/src/urma/lib/urma/bond/bondp_types.h
@@ -231,6 +231,7 @@ typedef struct bondp_comp {
     bondp_comp_type_t comp_type;
     urma_ref_t use_cnt; /* Initialize to 0 */
     // send
+    bool modify_to_error;
     bondp_jfc_t *send_jfc;
     pthread_spinlock_t send_lock;
     bool valid[URMA_UBAGG_DEV_MAX_NUM];
diff --git a/src/urma/lib/urma/bond/bondp_wr_buf.c b/src/urma/lib/urma/bond/bondp_wr_buf.c
index d8818553..f1104741 100644
--- a/src/urma/lib/urma/bond/bondp_wr_buf.c
+++ b/src/urma/lib/urma/bond/bondp_wr_buf.c
@@ -51,6 +51,7 @@ void wr_buf_uninit(wr_buf_t *buf)
 
         if (entry_hdr->entry_type == WR_BUF_ENTRY_JFS) {
             jfs_wr_entry_t *entry = (jfs_wr_entry_t *)__wr_buf_idx(buf, idx);
+            convert_jfs_pwr_to_vwr_resend(&entry->wr, &entry->v_conn->target_vjetty->v_tjetty);
             release_vwr_use_cnt(&entry->wr);
             free_jfs_wr(&entry->wr);
         } else if (entry_hdr->entry_type == WR_BUF_ENTRY_JFR) {
-- 
2.28.0.windows.1

